crates/sputnik/src/session.rs

```diff
@@ -162,3 +174,136 @@ fn get_repo_hash() -> Option<String> {
+            reporting_info: ReportingInfo {
+                is_telemetry_enabled: true,
+                endpoint: Url::parse(format!("http://0.0.0.0/{}", report_path()).as_str()).unwrap(),
+                user_agent: user_agent().into(),
+            },
+            client: Client::new(),
+        }
+    }
+
+    enum ReportCase {
+        Success,
+        TelemetryDisabled,
+        TimedOut,
+    }
+
+    #[rstest]
+    #[case::success(ReportCase::Success)]
+    #[case::telemetry_disabled(ReportCase::TelemetryDisabled)]
+    #[case::timedout(ReportCase::TimedOut)]
+    fn test_report(
+        #[case] case: ReportCase,
+        mut session: Session,
+        report_path: &'static str,
+        user_agent: &'static str,
+    ) -> Result<(), anyhow::Error> {
+        // Toggle between true/false for telemetry to test whether we fire requests when it's
+        // disabled
+        if let ReportCase::TelemetryDisabled = case {
+            session.reporting_info.is_telemetry_enabled = false;
+        } else {
+            session.reporting_info.is_telemetry_enabled = true;
+        }
+
+        let server = MockServer::start();
+        let addr = server.address().to_string();
+        let mocked_addr = Url::parse(format!("http://{}/{}", &addr, report_path).as_str()).unwrap();
+        session.reporting_info.endpoint = mocked_addr;
+
+        let mocked = server.mock(|when, then| {
+            when.method(POST)
+                // Annoyingly, the fixture needs to not have the preceding `/` for
+                // Url::parse() because it doesn't strip them; so, don't remove this
+                // preceding `/`
+                .path(format!("/{}", report_path))
+                .header("User-Agent", user_agent)
+                .header("Content-Type", "application/json");
+
+            match case {
+                ReportCase::Success | ReportCase::TelemetryDisabled => then.status(200),
+                ReportCase::TimedOut =>
+                // This won't actually wait 10s over the timeout threshold; the timeout
+                // will kick in and return an error
+                {
+                    then.status(504)
+                        .delay(REPORT_TIMEOUT + Duration::from_secs(10))
+                }
+            };
+        });
+
+        let res = session.report();
+
+        if let ReportCase::TelemetryDisabled = case {
+            // When telemetry is disabled, we should expect not outbound calls
+            mocked.assert_hits(0);
+        } else {
+            mocked.assert();
+        }
+
+        match case {
+            ReportCase::Success | ReportCase::TelemetryDisabled => {
+                assert_that!(res).is_ok();
+            }
+            ReportCase::TimedOut => {
+                assert_that!(res).is_err().matches(|err| {
+                    err.source().unwrap().source().unwrap().to_string() == "operation timed out"
+                });
+            }
+        };
+
+        Ok(())
+    }
+
+    //#[test]
```

Comment:
Looks like there are some extraneous bits down here.

Reply:
dang it, I thought I'd remover that